src/Storages/Statistics/Statistics.cpp

```diff
@@ -15,39 +18,134 @@ namespace ErrorCodes
+void ColumnStatistics::update(const ColumnPtr & column)
 {
-    if (!validators.emplace(stat_type, std::move(validator)).second)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "MergeTreeStatisticsFactory: the statistic validator type {} is not unique", stat_type);
+    rows += column->size();
+    for (const auto & iter : stats)
+    {
+        iter.second->update(column);
+    }
+}
 
+Float64 ColumnStatistics::estimateLess(Float64 val) const
+{
+    if (stats.contains(StatisticsType::TDigest))
+        return std::static_pointer_cast<TDigestStatistics>(stats.at(StatisticsType::TDigest))->estimateLess(val);
+    return rows * ConditionSelectivityEstimator::default_normal_cond_factor;
 }
 
-StatisticPtr TDigestCreator(const StatisticDescription & stat)
+Float64 ColumnStatistics::estimateGreater(Float64 val) const
 {
-    return StatisticPtr(new TDigestStatistic(stat));
+    return rows - estimateLess(val);
 }
 
-void TDigestValidator(const StatisticDescription &, DataTypePtr data_type)
+Float64 ColumnStatistics::estimateEqual(Float64 val) const
 {
-    data_type = removeNullable(data_type);
-    if (!data_type->isValueRepresentedByNumber())
-        throw Exception(ErrorCodes::ILLEGAL_STATISTIC, "TDigest does not support type {}", data_type->getName());
+    if (stats.contains(StatisticsType::Uniq) && stats.contains(StatisticsType::TDigest))
+    {
+        auto uniq_static = std::static_pointer_cast<UniqStatistics>(stats.at(StatisticsType::Uniq));
+        if (uniq_static->getCardinality() < 2048)
+        {
+            auto tdigest_static = std::static_pointer_cast<TDigestStatistics>(stats.at(StatisticsType::TDigest));
+            return tdigest_static->estimateEqual(val);
+        }
+    }
+    if (val < - ConditionSelectivityEstimator::threshold || val > ConditionSelectivityEstimator::threshold)
+        return rows * ConditionSelectivityEstimator::default_normal_cond_factor;
+    else
+        return rows * ConditionSelectivityEstimator::default_good_cond_factor;
 }
 
+void ColumnStatistics::serialize(WriteBuffer & buf)
+{
+    writeIntBinary(V0, buf);
+    UInt64 stat_types_mask = 0;
+    for (const auto & [type, _]: stats)
+    {
+        stat_types_mask |= 1 << UInt8(type);
+    }
+    writeIntBinary(stat_types_mask, buf);
+    /// We write some basic statistics
+    writeIntBinary(rows, buf);
+    /// We write complex statistics
+    for (const auto & [type, stat_ptr]: stats)
+    {
+        stat_ptr->serialize(buf);
+    }
+}
+
+void ColumnStatistics::deserialize(ReadBuffer &buf)
+{
+    UInt16 version;
+    readIntBinary(version, buf);
+    if (version != V0)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown file format version: {}", version);
+
+    UInt64 stat_types_mask = 0;
+    readIntBinary(stat_types_mask, buf);
+    readIntBinary(rows, buf);
+    for (auto it = stats.begin(); it != stats.end();)
+    {
+        if (!(stat_types_mask & 1 << UInt8(it->first)))
+        {
+            stats.erase(it ++);
```

Comment:
Cosmetics: Happy little whitespace errors in l. 105 and l. 110.

Reply:
For everyone's mental sanity, let's do one thing per line:

```cpp
stats.erase(it);
++it;
```

Reply:
It does not work for every container, because iterators could be invalidated.

You can do either:
```
it = stats.erase(it);
```
or
```
stats.erase(it++);
```
and both are idiomatic.